velox/expression/CastHooks.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/StringWriter.h"
+
+namespace facebook::velox::exec {
+
+/// This class provides cast hooks to allow different behaviors of CastExpr and
+/// SparkCastExpr. The main purpose is crate a Spark-compatible cast
+/// implementation by taking full usage of existing cast expression. The default
+/// logics are following Presto semantics.
+class CastHooks {
+ public:
+  virtual ~CastHooks() {}
+
+  // Returns a const reference of cast hooks.
+  static const std::shared_ptr<CastHooks>& get();
+
+  /// Cast from string to timestamp. TODO: The difference with Spark hook is
+  /// Spark allows more string patterns than Presto.
+  virtual Timestamp castStringToTimestamp(const StringView& view);
+
+  /// Returns the result of cast from string to date. The differences with Spark
+  /// hook are 1) When casting from string, Spark removes all leading and
+  /// trailing UTF8 white-spaces before cast. 2) The default implementation uses
+  /// standard cast mode, while Spark's hook uses non-standard one.
+  virtual int32_t castStringToDate(const StringView& dateString);
+
+  /// Cast from timestamp to string and write the result to string writer. The
+  /// differences with Spark hook are 1) the default implementation handles
+  /// legacy flag, while Spark's hook does not. 2) Spark's conversion is
+  /// in second-unit. 3) Spark adds a positive sign at first if the year exceeds
+  /// 9999.
+  virtual void castTimestampToString(
+      const Timestamp& timestamp,
+      bool legacy,
```

Comment:
The legacy flag is Presto-specific, right? It probably doesn't belong to this API. Currently, CastHooks is a singleton, but maybe it could be an instance created using query config. Then it can initialize itself as needed, e.g. set 'legacy' flag in Presto's implementation.

Reply:
Thanks, that makes sense. Changed to create hooks with config.